src/analysis/processing/qgsalgorithmlayoutatlastopdf.cpp

```diff
@@ -248,5 +256,182 @@ QVariantMap QgsLayoutAtlasToPdfAlgorithm::processAlgorithm( const QVariantMap &p
+      atlas->setSortFeatures( true );
+      atlas->setSortExpression( expression );
+      atlas->setSortAscending( !sortByReverse );
+    }
+    else
+    {
+      atlas->setSortFeatures( false );
+    }
+  }
+  else if ( !atlas->enabled() )
+  {
+    throw QgsProcessingException( QObject::tr( "Layout being export doesn't have an enabled atlas" ) );
+  }
+
+  const QgsLayoutExporter exporter( layout.get() );
+  QgsLayoutExporter::PdfExportSettings settings;
+
+  if ( parameters.value( QStringLiteral( "DPI" ) ).isValid() )
+  {
+    settings.dpi = parameterAsDouble( parameters, QStringLiteral( "DPI" ), context );
+  }
+  settings.forceVectorOutput = parameterAsBool( parameters, QStringLiteral( "FORCE_VECTOR" ), context );
+  settings.appendGeoreference = parameterAsBool( parameters, QStringLiteral( "GEOREFERENCE" ), context );
+  settings.exportMetadata = parameterAsBool( parameters, QStringLiteral( "INCLUDE_METADATA" ), context );
+  settings.simplifyGeometries = parameterAsBool( parameters, QStringLiteral( "SIMPLIFY" ), context );
+  settings.textRenderFormat = parameterAsEnum( parameters, QStringLiteral( "TEXT_FORMAT" ), context ) == 0 ? QgsRenderContext::TextFormatAlwaysOutlines : QgsRenderContext::TextFormatAlwaysText;
+
+  if ( parameterAsBool( parameters, QStringLiteral( "DISABLE_TILED" ), context ) )
+    settings.flags = settings.flags | QgsLayoutRenderContext::FlagDisableTiledRasterLayerRenders;
+  else
+    settings.flags = settings.flags & ~QgsLayoutRenderContext::FlagDisableTiledRasterLayerRenders;
+
+  settings.predefinedMapScales = QgsLayoutUtils::predefinedScales( layout.get() );
+
+  const QList< QgsMapLayer * > layers = parameterAsLayerList( parameters, QStringLiteral( "LAYERS" ), context );
+  if ( layers.size() > 0 )
+  {
+    const QList<QGraphicsItem *> items = layout->items();
+    for ( QGraphicsItem *graphicsItem : items )
+    {
+      QgsLayoutItem *item = dynamic_cast<QgsLayoutItem *>( graphicsItem );
+      QgsLayoutItemMap *map = dynamic_cast<QgsLayoutItemMap *>( item );
+      if ( map && !map->followVisibilityPreset() && !map->keepLayerSet() )
+      {
+        map->setKeepLayerSet( true );
+        map->setLayers( layers );
+      }
+    }
+  }
+  const QString filename = parameterAsString( parameters, QStringLiteral( "OUTPUT_FILENAME" ), context );
+
+  const QString destFolder = parameterAsFile( parameters, QStringLiteral( "OUTPUT_FOLDER" ), context );
+
+  const QString dest = QStringLiteral( "%1/%2.pdf" ).arg( destFolder, filename );
+
+  if ( atlas->updateFeatures() )
+  {
+    feedback->pushInfo( QObject::tr( "Exporting %n atlas feature(s)", "", atlas->count() ) );
+
+    QgsLayoutExporter::ExportResult result;
+    if ( !atlas->setFilenameExpression( filename, error ) )
+    {
+      throw QgsProcessingException( QObject::tr( "Output filename could not be set to create filenames for the files, \n"
+                                    "use a correct QGIS expression" ) );
+    }
```

Comment:
...and then handle a missing atlas filenameExpression
```suggestion
    if ( atlas->filenameExpression().isEmpty() && filename.isEmpty() )
    {
      atlas->setFilenameExpression( QString( "'output_'||@atlas_featurenumber" ), error );
    }
    else if ( !atlas->setFilenameExpression( filename, error ) )
    {
      throw QgsProcessingException( QObject::tr( "Output filename could not be set to create filenames for the files, \n"
                                    "use a correct QGIS expression" ) );
    }
```